### PR TITLE
vmware: Support cloud-init datasource GuestInfo

### PR DESCRIPTION
**Pull request by akutz**
**How to categorize this PR?**

/kind enhancement
/area os
/os garden-linux

**What this PR does / why we need it**:
This patch adds support for the Cloud-Init Datasource, VMwareGuestInfo, to the VMware image. The files in this commit were copied from release `v1.4.1` of the datasource from https://github.com/vmware/cloud-init-vmware-guestinfo.

The datasource has a dependency on the `netifaces` library, and so the package `python3-netifaces` was added to the package file for the `vmware` image.

The patch was verified by:

1. Building the VMware image:
    ```bash
    make vmware
    ```

2. Verifying the cloud-init datasource's files were copied into the image correctly:
    ```bash
    $ tar vtf .build/vmware/20210718/amd64/bullseye/rootfs.tar.xz | grep 'DataSourceVMware\|dscheck_VMware'
    -rw-rw-r-- 0/0             549 2021-07-18 00:00 etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg
    -rwxr-xr-x 0/0            2078 2021-07-18 00:00 usr/bin/dscheck_VMwareGuestInfo.sh
    -rw-rw-r-- 0/0           25937 2021-07-18 00:00 usr/lib/python3/dist-packages/cloudinit/sources/DataSourceVMwareGuestInfo.py
    ```

    Please note the file `dscheck_VMwareGuestInfo.sh` maintains its necessary, executable permissions.

3. Verifying the `python3-netifaces` package was installed correctly:
    ```bash
    $ tar vtf .build/vmware/20210718/amd64/bullseye/rootfs.tar.xz | grep 'netifaces'
    drwxr-xr-x 0/0               0 2021-07-18 00:00 usr/lib/python3/dist-packages/netifaces-0.10.9.egg-info/
    -rw-r--r-- 0/0           10291 2020-12-08 01:16 usr/lib/python3/dist-packages/netifaces-0.10.9.egg-info/PKG-INFO
    -rw-r--r-- 0/0               1 2020-12-08 01:16 usr/lib/python3/dist-packages/netifaces-0.10.9.egg-info/dependency_links.txt
    -rw-r--r-- 0/0              10 2020-12-08 01:16 usr/lib/python3/dist-packages/netifaces-0.10.9.egg-info/top_level.txt
    -rw-r--r-- 0/0               1 2018-05-02 09:15 usr/lib/python3/dist-packages/netifaces-0.10.9.egg-info/zip-safe
    -rw-r--r-- 0/0           23104 2020-12-08 01:16 usr/lib/python3/dist-packages/netifaces.cpython-39-x86_64-linux-gnu.so
    drwxr-xr-x 0/0               0 2021-07-18 00:00 usr/share/doc/python3-netifaces/
    -rw-r--r-- 0/0            1451 2018-04-02 20:08 usr/share/doc/python3-netifaces/copyright
    -rw-r--r-- 0/0             736 2021-07-18 00:00 var/lib/dpkg/info/python3-netifaces.list
    -rw-r--r-- 0/0             852 2020-12-08 01:16 var/lib/dpkg/info/python3-netifaces.md5sums
    ```

**Which issue(s) this PR fixes**:
Fixes NA

**Special notes for your reviewer**:
* This PR was initiated in conjunction with @marwinski .
* Please note this patch _could_ be implemented by leveraging `features/vmware/exec.confg` instead of copying static files. This is the process [used by the `kube-image-builder` project](https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/ansible/roles/providers/tasks/vmware.yml). However, a network-based, dynamic install is less deterministic than copying a few static files, so the latter method was used.
* Additionally, I initially had several issues building the image locally in a Linux VM. The root cause ended up being insufficient RAM for the VM. It had 4GiB, and I had to increase it to 8GiB for the image to build successfully.

**Release note**:

```feature operator
Support for the Cloud-Init datasource VMwareGuestInfo in the VMware image
```

### CODE REVIEW

**Comment by marwinski** [comment]
Thanks @akutz! Are you sure this does will break the existing vSphere 7.0 use case?

**Comment by akutz** [comment]
> Thanks @akutz! Are you sure this does will break the existing vSphere 7.0 use case?

Are you asking if I'm sure this won't break the existing 7.0 image?

If so, then no, I've not tested the image. You should do that yourself prior to merging this PR. What I am sure of is this PR solves the issue encountered with the cloud-init userdata example provided in that e-Mail thread.

File: features/vmware/file.include/etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+datasource_list: [ "VMwareGuestInfo" ]
```
Comment (marwinski) [inline features/vmware/file.include/etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg:13]: This breaks cloud-init for our existing landscape. This does fix it:

```
datasource_list: [ "VMwareGuestInfo", "OVF" ]
```

Can you please add this.

**Review by marwinski — CHANGES_REQUESTED** [review]
We haven't checked VMwareGuestInfo yet.

File: features/vmware/file.include/etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg (hunk 1 above)
Comment (MartinWeindel) [reply features/vmware/file.include/etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg:13]: After applying the change suggested by @marwinski, another problem occurs:
```
No check method 'dscheck_VMwareGuestInfo' for datasource 'VMwareGuestInfo'
```

The cause is the filename `features/vmware/file.include/usr/bin/dscheck_VMwareGuestInfo.sh`. It must be provided without the suffix `.sh`
see also https://github.com/canonical/cloud-init/blob/fbcb224bc12495ba200ab107246349d802c5d8e6/tools/ds-identify#L1686-L1687

Comment (MartinWeindel) [reply features/vmware/file.include/etc/cloud/cloud.cfg.d/99-DataSourceVMwareGuestInfo.cfg:13]: With both changes the image works both for our existing landscape and the staging vSphere in the lab.

**Review by marwinski — CHANGES_REQUESTED** [review]
we had to rename `features/vmware/file.include/usr/bin/dscheck_VMwareGuestInfo.sh`  to `features/vmware/file.include/usr/bin/dscheck_VMwareGuestInfo`. With this it worked for both environments.

**Comment by akutz** [comment]
> we had to rename `features/vmware/file.include/usr/bin/dscheck_VMwareGuestInfo.sh`  to `features/vmware/file.include/usr/bin/dscheck_VMwareGuestInfo`. With this it worked for both environments.

Ack. I'll submit the changes to the PR ASAP

**Comment by akutz** [comment]
Both changes are now applied. Please note that this datasource will become a standard part of cloud-init in 21.3 (https://github.com/canonical/cloud-init/pull/953).

**Comment by marwinski** [comment]
Hi @akutz. Thanks for the changes and sorry for the delay. We would like to include this in the upcoming Garden Linux release next week. Could you sign the contributor license agreement so that we can go ahead. Thanks.

**Comment by akutz** [comment]
> Hi @akutz. Thanks for the changes and sorry for the delay. We would like to include this in the upcoming Garden Linux release next week. Could you sign the contributor license agreement so that we can go ahead. Thanks.

Done and done!

**Review by marwinski — APPROVED** [review]

**Comment by akutz** [comment]
FYI @marwinski, I'm not sure on which Debian release Gardener Linux is based, but Debian [Bookworm](https://packages.debian.org/bookworm/cloud-init) and [Sid](https://packages.debian.org/sid/cloud-init) both have Cloud-Init v21.3. Since Bookworm is the next release for Debian, I imagine you will be able to take advantage of the [Cloud-Init DataSource for VMware](https://cloudinit.readthedocs.io/en/latest/topics/datasources/vmware.html) soon.

**Comment by akutz** [comment]
@marwinski The GitHub check above says I've not yet signed the CLA when I clearly have signed the CLA:

image

**Review by gehoern — APPROVED** [review]

**Pull request merged**